### PR TITLE
add watcher support for xds proxy certs

### PR DESCRIPTION
**Pull request by ramaraochavali**
- Adds watch support for xds proxy client certs and closes upstream connection when they change
- Handle file mounted certs

Still needs to add tests.

[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[ X] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ X] Does not have any changes that may affect Istio users.

### CODE REVIEW

File: pkg/istio-agent/xds-proxy.go
Comment (howardjohn) [inline pkg/istio-agent/xds-proxy.go:464]: is this not a constant?

File: pkg/istio-agent/xds-proxy.go
Comment (howardjohn) [inline pkg/istio-agent/xds-proxy.go:436]: I don't understand why we need to reset explicitly. I thought that GetClientCertificate is called on every new connection already? Even if the cert changes, our existing connection will have already done the TLS handshake and should be fine? This code seems to drift from CSR code

**Review by howardjohn — COMMENTED** [review]
I have some concerns about this. Part of the goal of this is to consolidate with the CSR stuff (in terms of code and physical connection). This seems to imply either:

* CSR does not properly watch certs, and we should fix it
* We are drifting further from CSR code.

I am ok with taking temporary tech debt and drifting from the code, but we should try to keep them as close as possible if we need to and clearly document this in the code IMO. Specifically I think the tls.Config is copy+pasted from CSR code previously and is not different which is concerning.

File: pkg/istio-agent/xds-proxy.go
Comment (ramaraochavali) [reply pkg/istio-agent/xds-proxy.go:436]: Yes. Existing connections is fine. The new connection might only happen on the next disconnect either forceful disconnect by Istiod or because of some network disconnect. Assuming that we want to do intelligent load balancing at proxy in future, the forceful disconnect might not be there. 

So do we want the new certs only to be picked on new connections (which is not in our control when they might happen)?  Comparing with today's functionality with out xds proxy, the certs would be reloaded when they change via SDS? I am trying to match that behaviour here.

But if we think, it is better to not watch and let the new certs take affect only on new connection - I am fine with that.

**Comment by ramaraochavali** [comment]
So your concern is more about adding watcher support? Can you point me to CSR code?

**Comment by howardjohn** [comment]
I'm not necessarily concerned with add watcher just want to make sure it's consistent with csr. On a phone now so cannot find that code.

What does envoy do when we push a new secret? Will it forcefully reconnect?

**Comment by ramaraochavali** [comment]
> What does envoy do when we push a new secret? Will it forcefully reconnect?

If the secret is pushed for listener, it will drain the listener.  If the secret is pushed for a cluster , after cluster is warmed up it should drain all previous connections.

File: pkg/istio-agent/xds-proxy.go
Comment (rshriram) [inline pkg/istio-agent/xds-proxy.go:235]: i just remembered. do we need to do something to make sure the go routines in this func terminate?

File: pkg/istio-agent/xds-proxy.go
Comment (rshriram) [inline pkg/istio-agent/xds-proxy.go:388]: need checks to see if file certs are used, and if not return nil, without setting up go routine

File: pkg/istio-agent/xds-proxy.go
Comment (ramaraochavali) [reply pkg/istio-agent/xds-proxy.go:235]: upstream go routine will go away - downstream one we need to see, I will take a look.

File: pkg/istio-agent/xds-proxy.go
Comment (ramaraochavali) [reply pkg/istio-agent/xds-proxy.go:468]: No. There are no constants for them.

**Review by rshriram — COMMENTED** [review]
lgtm from me as long as the duplication goes away soon. @howardjohn ?

File: pkg/istio-agent/xds-proxy.go
Comment (howardjohn) [reply pkg/istio-agent/xds-proxy.go:394]: Not sure if this is done or not - but the check on L397 doesn't quite do this, we will always have a root cert, so it will always be used. Maybe that is a good idea though

File: pkg/istio-agent/xds-proxy.go
Comment (howardjohn) [inline pkg/istio-agent/xds-proxy.go:476]: why doesn't FindRootCAForXDS encode the `MetadataClientRootCert` logic? The whole point of the FindRootCAForXDS is to avoid this type of logic proliferation

File: pkg/istio-agent/xds-proxy.go
Comment (howardjohn) [inline pkg/istio-agent/xds-proxy.go:64]: are these used *only* for XDS? Or for CSR as well?

File: pkg/istio-agent/xds-proxy.go
Comment (ramaraochavali) [reply pkg/istio-agent/xds-proxy.go:476]: Sure. It can. I will make that change.

File: pkg/istio-agent/xds-proxy.go
Comment (ramaraochavali) [reply pkg/istio-agent/xds-proxy.go:64]: It is not used for CSR - BTW, can you point me to that code - I am not aware of that code

Comment (howardjohn) [reply pkg/istio-agent/xds-proxy.go:64]: https://github.com/istio/istio/blob/2c81d37a8627ed2b880ee8dd961f7982b94ab1d2/security/pkg/nodeagent/caclient/providers/citadel/client.go#L137

File: pkg/istio-agent/xds-proxy.go
Comment (ramaraochavali) [reply pkg/istio-agent/xds-proxy.go:394]: I did not know that root will always be there. I have corrected that check

**Comment by ramaraochavali** [comment]
@howardjohn addressed comments. PTAL

File: pkg/istio-agent/agent.go
Comment (howardjohn) [inline pkg/istio-agent/agent.go:338]: is FileMountedCerts ever set when MetadataClientRootCert is not? let me look

Comment (ramaraochavali) [reply pkg/istio-agent/agent.go:338]: We do not have that validation in-place but it never happens

Comment (ramaraochavali) [reply pkg/istio-agent/agent.go:338]: are you thinking it would be safe to check existence of that certs vs isFileMountedcerts? IF yes, I am OK moving to that check

Comment (howardjohn) [reply pkg/istio-agent/agent.go:338]: My main concern is if somehow this value is set but then the root cert path wasn't. That seems like it's probably a misconfiguration but possibly one that's easy to misconfigured? So maybe we just need to have some log or failure in that case

File: pkg/istio-agent/xds-proxy.go
Comment (ramaraochavali) [reply pkg/istio-agent/xds-proxy.go:256]: If wer return here, it will restart the whole process - right?

Comment (howardjohn) [reply pkg/istio-agent/xds-proxy.go:256]: missed that, lgtm

File: pkg/istio-agent/agent.go
Comment (ramaraochavali) [reply pkg/istio-agent/agent.go:338]: `getRootCertificate` already errors out if the file determined  by FindRootCAForXDS does not exists. That should cover this case?

**Review by howardjohn — APPROVED** [review]

**Comment by costinm** [comment]
I am very concerned with the divergence from the SDS code paths - we finally managed to consolidate all cert-related logic to SDS, and now we're back to watching files ? 

Also once again - please merge the xds proxy with adsc, so we can share all the auth-related logic ( JWT, etc).

**Comment by ramaraochavali** [comment]
> I am very concerned with the divergence from the SDS code paths - we finally managed to consolidate all cert-related logic to SDS, and now we're back to watching files ?

This is for xds proxy client certs to Istiod. There is no SDS involved here.

**Comment by costinm** [comment]
We don't have to use SDS for the workload cert - ok to call directly into the SecretCache. In Envoy the XDS connection uses a cert retrieved from SDS. 

The logic around certs is extremely complicated with too many options ( most of them unfortunately needed because they reflect real use cases and customer needs ). We now have 3 variants ( envoy, xds proxy and adsc/istioclt ) each with a different code path.

**Comment by costinm** [comment]
Also: if SDS is not involved then it's clearly broken for GKE - since the integration of Envoy with GKE is in SDS code ( SecretCache) ( I'm talking about using Istio with Google Managed CA, which is used by some GKE users )

**Pull request merged**
